app/src/test/java/org/simple/clinic/home/overdue/phonemask/PhoneMaskBottomSheetControllerTest.kt

```diff
@@ -0,0 +1,87 @@
+package org.simple.clinic.home.overdue.phonemask
+
+import com.nhaarman.mockito_kotlin.mock
+import com.nhaarman.mockito_kotlin.verify
+import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
+import com.nhaarman.mockito_kotlin.whenever
+import io.reactivex.Completable
+import io.reactivex.subjects.PublishSubject
+import junitparams.JUnitParamsRunner
+import junitparams.Parameters
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.simple.clinic.home.overdue.phonemask.CallClicked.NormalCallClicked
+import org.simple.clinic.home.overdue.phonemask.CallClicked.SecureCallClicked
+import org.simple.clinic.patient.PatientMocker.overduePatient
+import org.simple.clinic.phone.Caller
+import org.simple.clinic.phone.Caller.UsingDialer
+import org.simple.clinic.phone.Caller.WithoutDialer
+import org.simple.clinic.phone.MaskedPhoneCaller
+import org.simple.clinic.util.RuntimePermissionResult
+import org.simple.clinic.util.RuntimePermissionResult.DENIED
+import org.simple.clinic.util.RuntimePermissionResult.GRANTED
+import org.simple.clinic.util.RuntimePermissionResult.NEVER_ASK_AGAIN
+import org.simple.clinic.util.RxErrorsRule
+import org.simple.clinic.widgets.UiEvent
+
+@RunWith(JUnitParamsRunner::class)
+class PhoneMaskBottomSheetControllerTest {
+
+  @get:Rule
+  val rxErrorsRule = RxErrorsRule()
+
+  private val screen = mock<PhoneMaskBottomSheet>()
+  private val uiEvents = PublishSubject.create<UiEvent>()
+  private val phoneCaller = mock<MaskedPhoneCaller>()
+  private val controller = PhoneMaskBottomSheetController(phoneCaller)
+
+  @Before
+  fun setUp() {
+    uiEvents.compose(controller).subscribe { uiChange -> uiChange(screen) }
+  }
+
+  @Test
+  @Parameters(method = "params for types of call")
+  fun `when any call button is clicked, call permission should be requested`(callTypeEvent: CallClicked) {
+    uiEvents.onNext(PhoneMaskBottomSheetCreated(overduePatient(phoneNumber = "1234567890")))
+    uiEvents.onNext(callTypeEvent)
+
+    verify(screen).requestCallPermission()
+    verifyNoMoreInteractions(screen)
+  }
+
+  private fun `params for types of call`() =
+      listOf(NormalCallClicked, SecureCallClicked)
+
+  @Test
+  @Parameters(method = "params for making phone calls")
+  fun `when any call button is clicked and permission result is received, appropriate call should be made`(
+      callTypeEvent: CallClicked,
+      permission: RuntimePermissionResult,
+      caller: Caller
+  ) {
+    val number = "1234567890"
+
+    whenever(phoneCaller.normalCall(number, caller)).thenReturn(Completable.complete())
+    whenever(phoneCaller.maskedCall(number, caller)).thenReturn(Completable.complete())
+
+    uiEvents.onNext(PhoneMaskBottomSheetCreated(overduePatient(phoneNumber = number)))
+    uiEvents.onNext(callTypeEvent)
+    uiEvents.onNext(CallPhonePermissionChanged(permission))
+
+    verify(screen).requestCallPermission()
+    verifyNoMoreInteractions(screen)
```

Comment:
This test does not verify that the callers are invoked, nor that the Completables that they emit are subscribed to.

I usually tend to set a flag externally from the `doOnSubscribe` side effect block of any Completables that I return to verify this.

Reply:
done